### PR TITLE
Azure Storage:  Azure storage support #3253

### PR DESCRIPTION
**Pull request by sfali**
References #3253 

Work in progress, still needs to add more tests and documentations.

Code style is copied from AWS S3 connector.

### CODE REVIEW

**Comment by sfali** [comment]
This PR is ready to be reviewed,  documentation pipeline is failing but can't figure why.

Few mock test cases are ignored the reason being that Azure expect to have `Content-Length` header to be set explicitly even the length is `0`. `Akka HTTP` doesn't add this header, which makes request fails. But by adding this header, it seems WireMock also add `Content-Length` header of its own which makes two `Content-Length` headers in the request and test fails because of this. Firstly need to figure out possibility of WireMock not to add `Content-Length` header.

Even if I comment out explicit addition of `Content-Length` header, test fails with `akka.http.scaladsl.model.EntityStreamException`, not sure why?

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+import akka.stream.{Attributes, Materializer}
+import akka.stream.alpakka.azure.storage.impl.auth.Signer
+import akka.stream.scaladsl.{Flow, RetryFlow, Source}
+import akka.util.ByteString
+
+import scala.collection.immutable
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+object AzureStorageStream {
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala:39]: Implementation clas

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/javadsl/BlobService.scala [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+import akka.stream.scaladsl.SourceToCompletionStage
+import akka.util.ByteString
+
+import java.util.Optional
+import java.util.concurrent.CompletionStage
+
+/**
+ * Java API for BlobService operations.
+ */
+object BlobService {
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/javadsl/BlobService.scala:27]: Java DSL

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/javadsl/FileService.scala [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+import akka.stream.scaladsl.SourceToCompletionStage
+import akka.util.ByteString
+
+import java.util.Optional
+import java.util.concurrent.CompletionStage
+
+/**
+ * Java API FileService operations
+ */
+object FileService {
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/javadsl/FileService.scala:28]: Java DSL for FileService operations

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/scaladsl/BlobService.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+import akka.stream.alpakka.azure.storage.impl.AzureStorageStream
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+
+import scala.concurrent.Future
+
+/**
+ * Scala API for BlobService operations.
+ */
+object BlobService {
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/scaladsl/BlobService.scala:23]: Scala DSL for BlobService operations

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/scaladsl/FileService.scala [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import akka.stream.alpakka.azure.storage.impl.AzureStorageStream
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+
+import scala.concurrent.Future
+
+/**
+ * Scala API for FileService operations.
+ */
+object FileService {
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/scaladsl/FileService.scala:23]: Scala DSL for FileService operations

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/requests/RequestBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+package azure
+package storage
+package requests
+
+import akka.http.scaladsl.model.Uri.Query
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{HttpHeader, HttpMethod, HttpRequest, Uri}
+import akka.stream.alpakka.azure.storage.headers.ServerSideEncryption
+
+abstract class RequestBuilder(val sse: Option[ServerSideEncryption] = None,
```
Comment (sfali) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/requests/RequestBuilder.scala:15]: Base class for request builder

File: azure-storage/src/main/resources/reference.conf [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      account-key = none
+      account-key = ${?AZURE_STORAGE_ACCOUNT_KEY}
+
+      # SAS token for sas authorization
+      sas-token = ""
+      sas-token = ${?AZURE_STORAGE_SAS_TOKEN}
+    }
+    #azure-credentials
+
+    # Default settings corresponding to automatic retry of requests in an S3 stream.
```
Comment (girdharshubham) [inline azure-storage/src/main/resources/reference.conf:30]:
```suggestion
    # Default settings corresponding to automatic retry of requests in an Azure Storage stream.
```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import akka.stream.Attributes
+import akka.stream.Attributes.Attribute
+
+/**
+ * Akka Stream attributes that are used when materializing AzureStorage stream blueprints.
+ */
+object StorageAttributes {
+
+  /**
+   * Settings to use for the S3 stream
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala:18]:
```suggestion
   * Settings to use for the Azure Storage stream
```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+      .parseFromValueString(entity.contentType.value)
+      .map(Seq(_))
+      .getOrElse(Nil)
+    ObjectMetadata(
+      headers ++ contentLengthHeader ++ contentTypeHeader ++ Seq(CustomContentTypeHeader(entity.contentType))
+    )
+  }
+
+  private def isTransientError(status: StatusCode): Boolean = {
+    // 5xx errors from S3 can be treated as transient.
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala:220]:
```suggestion
    // 5xx errors from Azure Storage can be treated as transient.
```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/models.scala [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    AzureNameKeyCredential(accountName, accountKey)
+
+  def apply(config: Config): AzureNameKeyCredential = {
+    val accountName = config.getString("account-name", "")
+    val accountKey = config.getString("account-key", "")
+    AzureNameKeyCredential(accountName, accountKey)
+  }
+}
+
+/** Modelled after ObjectMetadata in S3.
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/models.scala:34]:
```suggestion
/** Modelled after BlobProperties in Azure Blob Storage.
```

File: azure-storage/src/test/resources/logback-test.xml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+<configuration>
+    <appender name="FILE" class="ch.qos.logback.core.FileAppender">
+        <file>target/s3.log</file>
```
Comment (girdharshubham) [inline azure-storage/src/test/resources/logback-test.xml:3]:
```suggestion
        <file>target/azure-blob-storage.log</file>
```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+   */
+  def settingsPath(path: String): Attributes = Attributes(StorageSettingsPath(path))
+
+  /**
+   * Default settings
+   */
+  def defaultSettings: Attributes = Attributes(StorageSettingsPath.Default)
+}
+
+final class StorageSettingsPath private (val path: String) extends Attribute
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala:33]:
```suggestion
final class StorageSettingsPath private (val path: String) extends Attribute

```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+object StorageSettingsPath {
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala:34]:
```suggestion

object StorageSettingsPath {
```

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  val Default: StorageSettingsPath = StorageSettingsPath(StorageSettings.ConfigPath)
+
+  def apply(path: String) = new StorageSettingsPath(path)
+}
+
+final class StorageSettingsValue private (val settings: StorageSettings) extends Attribute
+object StorageSettingsValue {
```
Comment (girdharshubham) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/StorageAttributes.scala:41]:
```suggestion

object StorageSettingsValue {
```

**Review by girdharshubham — COMMENTED** [review]
Looking good so far.

**Comment by sfali** [comment]
Issue with Java API was misconfiguration, not sure why `AzuriteIntegrationSpec` is failing, any idea @girdharshubham?

**Comment by girdharshubham** [comment]
> Issue with Java API was misconfiguration, not sure why `AzuriteIntegrationSpec` is failing, any idea @girdharshubham?

Hi @sfali 
Sorry this got delayed, I will check this soon.

**Comment by sebastian-alfers** [comment]
Hi @sfali, can you share and update on this PR? Do you want to see your contribution merged for the next release?

**Comment by sfali** [comment]
> Hi @sfali, can you share and update on this PR? Do you want to see your contribution merged for the next release?

What do you want me to share @sebastian-alfers? The PR is failing due to `couchbase` and `ironmq` connectors are failing. not sure what should I do with these connectors.

**Comment by sfali** [comment]
> We fixed both in `main`, so feel free to rebase and give it a try.

@sebastian-alfers Now scaaldoc and jms connector failed. Can pin point what's going? Can you please take a look and let me know what's going on.

Thanks

**Comment by sfali** [comment]
> Thank you for bringing this to our attention. The scaladoc failure can be ignored for now, and the jms was something we just fixed in main, so please rebase again.

Thanks @sebastian-alfers , Everything is passing other than scaaldoc

**Comment by sebastian-alfers** [comment]
Please let me know once you want to merge it.

**Comment by sfali** [comment]
> Please let me know once you want to merge it.

This can be merge anytime. It needs to be approved.

**Review by sebastian-alfers — APPROVED** [review]

**Comment by sfali** [comment]
It seems scaladoc pipeline needs to be successful before it can be merged.

File: azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+                  requestBuilder = requestBuilder)
+
+  private[storage] def deleteFile(objectPath: String,
+                                  requestBuilder: RequestBuilder): Source[Option[ObjectMetadata], NotUsed] =
+    handleRequest(successCode = Accepted,
+                  storageType = FileType,
+                  objectPath = objectPath,
+                  requestBuilder = requestBuilder)
+
+  private[storage] def putBlob(objectPath: String,
```
Comment (Susmit07) [inline azure-storage/src/main/scala/akka/stream/alpakka/azure/storage/impl/AzureStorageStream.scala:96]: @sfali one question for large files may be of > 5-10Gb .. S3 has multipartUploadAPI for azure blob how are we going to achieve it ?

splitting in chunks? 

adding some UTs would have been helpful..

**Pull request merged**